lib/mayaUsdAPI/render.h

```diff
@@ -105,20 +108,59 @@ class MAYAUSD_API_PUBLIC OgsFragment
 
 namespace ShaderGenUtil {
 
-// Forward declarations.
+// Forward declaration:
+struct LobePrunerImpl;
 struct TopoNeutralGraphImpl;
 
+/*! \brief  This class will process MaterialX surface shader nodes and provide optimized version of
+ * the shader based on the current value of a node attribute.
+ */
+class MAYAUSD_API_PUBLIC LobePruner
+{
+public:
+    using Ptr = std::shared_ptr<LobePruner>;
+
+    /*! Creates a shared LobePruner.
+     *  \return the freshly created LobePruner.
+     */
+    static Ptr create();
+
+    ~LobePruner();
+
+    /*! Sets the LobePruner library. This is used first to explore all surface shaders for
+     * optimization candidated, then to store the optimized NodeDef and NodeGraph this class
+     * generates.
+     * @param[in] library is a fully loaded MaterialX library where the generated optimized shaders
+     * will be stored.
+     */
+    void setLibrary(const MaterialX::DocumentPtr& library);
+
+private:
+    LobePruner();
+
+    std::unique_ptr<LobePrunerImpl> _imp;
+    friend struct TopoNeutralGraphImpl;
+};
+
 class MAYAUSD_API_PUBLIC TopoNeutralGraph
 {
 public:
     TopoNeutralGraph(const MaterialX::ElementPtr&);
+    TopoNeutralGraph(const MaterialX::ElementPtr&, const LobePruner::Ptr& lobePruner);
     ~TopoNeutralGraph();
 
     MaterialX::NodeGraphPtr nodeGraph();
 
     MaterialX::DocumentPtr getDocument() const;
     const std::string&     getOriginalPath(const std::string& topoPath) const;
 
+    /*! Get the list of node.attribute paths used by the LobePruner to optimize surface shader nodes
+     * found in the material that was processed.
+     * \return a string vector containing all paths to attibutes that were used to optimize nodes in
+     * this graph.
+     */
+    const MaterialX::StringVec& getOptimizedAttributes() const;
```

Comment:
Same here. Not virtual. And since the actual implementation is hidden underneath an Impl pointer, all changes to the core class are unexposed and non-breaking.